RemoveDuplicateVertices Implementation for TriangleMesh.

intelshashi opened the pull request:
Functionality to remove duplicate vertices and update other vertex attributes and triangle indices is implemented. The C++ and python tests are also written to test the code.

On branch sganjugu/remdup2
Changes to be committed:
modified:   cpp/open3d/t/geometry/TriangleMesh.cpp
modified:   cpp/open3d/t/geometry/TriangleMesh.h
modified:   cpp/pybind/t/geometry/trianglemesh.cpp
modified:   cpp/tests/t/geometry/TriangleMesh.cpp
modified:   python/test/t/geometry/test_trianglemesh.py

<!--- Provide a general summary of your changes in the Title above -->

## Type

<!--- Select with 'x' and link to a related issue. What types of changes does your code introduce? -->

-   [ ] Bug fix (non-breaking change which fixes an issue): Fixes #
-   [x] New feature (non-breaking change which adds functionality). Resolves #
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) Resolves #

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that
apply.  If you're unsure about any of these, don't hesitate to ask. We're here
to help! -->

-   [ ] I have run `python util/check_style.py --apply` to apply Open3D **code style**
    to my code.
-   [x] This PR changes Open3D behavior or adds new functionality.
    -   [ ] Both C++ (Doxygen) and Python (Sphinx / Google style) **documentation** is
        updated accordingly.
    -   [ ] I have added or updated C++ and / or Python **unit tests** OR included **test
        results** (e.g. screenshots or numbers) here.
-   [ ] I will follow up and update the code if CI fails.
    <!-- In case I am unavailable later -->
-   [x] For fork PRs, I have selected **Allow edits from maintainers**.

## Description

<!--- Describe your changes, with test results and screenshots as appropriate. Move unrelated changes, if any, to a separate PR. -->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/isl-org/Open3D/6414)
<!-- Reviewable:end -->

intelshashi committed:
RemoveDuplicateVertices Implementation for TriangleMesh.

Functionality to remove duplicate vertices and update
other vertex attributes and triangle indices is implemented.
The C++ and python tests are also written to test the code.

On branch sganjugu/remdup2
Changes to be committed:
modified:   cpp/open3d/t/geometry/TriangleMesh.cpp
modified:   cpp/open3d/t/geometry/TriangleMesh.h
modified:   cpp/pybind/t/geometry/trianglemesh.cpp
modified:   cpp/tests/t/geometry/TriangleMesh.cpp
modified:   python/test/t/geometry/test_trianglemesh.py

intelshashi committed:
Updated CHANGELOG.md

ssheorey commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1061,6 +1061,188 @@ TriangleMesh TriangleMesh::SelectFacesByMask(const core::Tensor &mask) const {
+/// \param attrib The attribute whose tensor has to be shrunk.
+/// \param mesh The mesh from which duplicates have to be removed.
+/// \param new_size The size to shrink the attribute tensor to.
+void ShrinkVertexAttributeTensor(const std::string &attrib, TriangleMesh &mesh, int64_t new_size)    
```
Put privately used functions into an anonmouse namespace (or make them static) (private Open3D symbol should not leak outside libOpen3D).

ssheorey commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1061,6 +1061,188 @@ TriangleMesh TriangleMesh::SelectFacesByMask(const core::Tensor &mask) const {
+/// \param index_old_to_new Map from old indices to new indices.
+/// \param mesh The mesh on which the new index mapping for triangles is to be computed
+template<typename T, typename Length_t>
+void RemapTriangleIndices(T *indices_ptr, std::vector<Length_t> &index_old_to_new, TriangleMesh &mesh)
```
anon namespace

intelshashi replied:
This function is now removed.

ssheorey commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1061,6 +1061,188 @@ TriangleMesh TriangleMesh::SelectFacesByMask(const core::Tensor &mask) const {
+///   which can do a better job of comparing the double values.
+/// \param mesh The mesh from which duplicate vertices are to be removed.
+template<typename Coord_t, typename Tindex_t>
+TriangleMesh& RemoveDuplicateVerticesWorker(TriangleMesh &mesh)
```
anon namespace

intelshashi replied:
Fixed.

ssheorey commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1061,6 +1061,188 @@ TriangleMesh TriangleMesh::SelectFacesByMask(const core::Tensor &mask) const {
+{
+    auto old_tensor = mesh.GetVertexAttr(attrib);
+    std::vector<decltype(new_size)> old_shape_vec(old_tensor.GetShape());
+    assert (new_size <= old_shape_vec[0]);
```
Use `utility::LogError` instead of `assert` for runtime checks to allow runtime user control over checks.
For debug only checks use OPEN3D_ASSERT macro

ssheorey reviewed (changes requested):
Hi @intelshashi , this approach and implementation are sound and logical. Since Open3D needs to be performance portable on both CPU and GPU, we have added some optimized Tensor primitives to ease algorithm development. You can see some of these in `TriangleMesh::SelectFacesByMask` which solves a similar problem - keeping a subset of vertices. That approach ensures that these operations work on both CPU and GPU and are performant on the GPU. I would recommend switching to that approach before we merge this PR. Specifically:
- create a `vertex_mask` of vertices to be kept and use `Tensor::IndexGet` to subset vertices and attributes.
- Use `utility::InclusivePrefixSum` to remap vertex indices.

intelshashi committed:
Moving standalone functions to anonymous namespace

intelshashi committed:
Duplicate vertex attributes shrunk using IndexGet

Previously, I was doing a manual copy and shrinking the vertices.
Instead in this checkin all vertex attributes, including the
coordinates are shrunk using IndexGet method using a vertex mask.

Further, the triangle index remapping is similar to what was done
earlier, with some simplications. One thing to note, is that
we cannot use utility::InclusivePrefixSum to remap vertex indices
because the duplicates can occur in any position and may be duplicates
of any earlier vertex.

For e.g., suppose there were 9 vertices to start with, and 8th (index 7, starting from 0),
was a duplicate of 2nd (index 1, starting from 0).

So, the vertex mask would look like this:
Vertex indices: [0, 1, 2, 3, 4, 5, 6, 7, 8]
Vertex mask:    [1, 1, 1, 1, 1, 1, 1, 0, 1]
Prefix sum:   [0, 1, 2, 3, 4, 5, 6, 7, 7, 8]
This gives an incorrect index map for 8th vertex, which is mapped to
index 7 instead of 1.

On branch sganjugu/remdup2
Your branch is up to date with 'origin/sganjugu/remdup2'.
Changes to be committed:
modified:   ../cpp/open3d/t/geometry/TriangleMesh.cpp
modified:   ../python/test/t/geometry/test_trianglemesh.py

intelshashi commented:
> Hi @intelshashi , this approach and implementation are sound and logical. Since Open3D needs to be performance portable on both CPU and GPU, we have added some optimized Tensor primitives to ease algorithm development. You can see some of these in `TriangleMesh::SelectFacesByMask` which solves a similar problem - keeping a subset of vertices. That approach ensures that these operations work on both CPU and GPU and are performant on the GPU. I would recommend switching to that approach before we merge this PR. Specifically:
> 
> * create a `vertex_mask` of vertices to be kept and use `Tensor::IndexGet` to subset vertices and attributes.
> * Use `utility::InclusivePrefixSum` to remap vertex indices.

intelshashi commented:
> > Hi @intelshashi , this approach and implementation are sound and logical. Since Open3D needs to be performance portable on both CPU and GPU, we have added some optimized Tensor primitives to ease algorithm development. You can see some of these in `TriangleMesh::SelectFacesByMask` which solves a similar problem - keeping a subset of vertices. That approach ensures that these operations work on both CPU and GPU and are performant on the GPU. I would recommend switching to that approach before we merge this PR. Specifically:
> > 
> > * create a `vertex_mask` of vertices to be kept and use `Tensor::IndexGet` to subset vertices and attributes.
> > * Use `utility::InclusivePrefixSum` to remap vertex indices.

For e.g., suppose, you had 9 vertices, and 8th (index 7, starting from 0), was a duplicate of 2nd (index 1, starting from 0).
So, the vertex mask would look like this:
Vertex indices: [0, 1, 2, 3, 4, 5, 6, 7, 8]
Vertex mask:    [1, 1, 1, 1, 1, 1, 1, 0, 1]
Prefix sum:   [0, 1, 2, 3, 4, 5, 6, 7, 7, 8]
This gives an incorrect index map for 8th vertex.
In fact it could be a duplicate of any of the previous vertices.

intelshashi committed:
Merge branch 'master' into sganjugu/remdup2

nsaiapova commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1215,6 +1215,139 @@ TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+    auto orig_num_vertices = vertices.GetLength();
+    const auto vmask_type = core::Int32;
+    using vmask_itype = int32_t;
+    core::Tensor vertex_mask = core::Tensor::Zeros({orig_num_vertices}, vmask_type);
```
If the mask is not used to store the remapped indices, I think we can have it as Bool right from the initialization.

nsaiapova commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1215,6 +1215,139 @@ TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+    auto triangles = GetTriangleIndices();
+    if (core::Int32 != triangles.GetDtype() && core::Int64 != triangles.GetDtype()) {
+        utility::LogError("Only Int32 or Int64 are supported for triangle indices");
+        return *this;
```
I think we don't need to return after `utility::LogError`. (here and below)

nsaiapova commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1215,6 +1215,139 @@ TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+        return *this;        
+    }
+    
+    if(core::Float32 == vertices.GetDtype()) {
+        if (core::Int32 == triangles.GetDtype()) {
+            return RemoveDuplicateVerticesWorker<float, int>(*this);
+        } else {
+            return RemoveDuplicateVerticesWorker<float, int64_t>(*this);
+        }
+    } else {
+        if (core::Int32 == triangles.GetDtype()) {
+            return RemoveDuplicateVerticesWorker<double, int>(*this);
+        } else {
+            return RemoveDuplicateVerticesWorker<double, int64_t>(*this);
+        }
+    }
```
While working on my `SelectByIndex` method, I discovered a set of macros to do this in more compact way:
https://github.com/isl-org/Open3D/blob/main/cpp/open3d/core/Dispatch.h#L90
I think we could rewrite it as smt like:
```suggestion
    DISPATCH_FLOAT_INT_DTYPE_TO_TEMPLATE(vertices.GetDtype(), triangles.GetDtype()), [&]() {
        RemoveDuplicateVerticesWorker<scalar_t, int_t>(*this);
    });
```

nsaiapova commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1215,6 +1215,139 @@ TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+TriangleMesh& RemoveDuplicateVerticesWorker(TriangleMesh &mesh)
+{   
+    //Algorithm based on Eigen implementation 
+    if(!mesh.HasVertexPositions()) {
```
Could you please run `python util/check_style.py --apply` on your changes? There are a few accidental space misses.

nsaiapova commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1215,6 +1215,139 @@ TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+    //REM_DUP_VERT_STEP 1: 
+    // Compute map from points to old indices, and use a counter
+    // to compute a map from old indices to new unique indices.
+    const Coord_t *vertices_ptr = vertices.GetDataPtr<Coord_t>();
```
I think we need to move `vertices` tensor to CPU device, too.

nsaiapova commented on cpp/open3d/t/geometry/TriangleMesh.cpp:
```diff
@@ -1215,6 +1215,139 @@ TriangleMesh TriangleMesh::SelectByIndex(const core::Tensor &indices) const {
+        return *this;
+    }
+    auto vertices = GetVertexPositions();
+    auto triangles = GetTriangleIndices();
```
There could be meshes w/o triangles tensor. In this case, the call to `GetTrianglesIndices` throws.
Could you please add safety checks for this case?

ssheorey commented:
Close in favor of #6636
